Document complex optimizer semantic behavior

janeyx99 opened the pull request:
<img width="817" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/565b389d-3e86-4767-9fcb-fe075b50aefe">

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121667

janeyx99 committed:
Document complex optimizer semantic behavior

[ghstack-poisoned]

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -129,9 +129,33 @@ Autograd
+`real_optim` and `complex_optim` will compute the same updates on the parameters. Note that the :func:`torch.view_as_real`
+equivalent will convert a complex tensor to a real tensor with shape :math:`(..., 2)`, which is distinct from splitting
+every complex parameter into `p.real` and `p.imag` and feeding twice the number of parameters into the optimizer. This
+distinction is trivial for pointwise optimizers (like AdamW) but will cause slight discrepancy in second order optimizers
```
nit: does "trivial" here as the meaning of "no impact"? I think we should use a clearer word.

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -129,9 +129,33 @@ Autograd
+
+
+`real_optim` and `complex_optim` will compute the same updates on the parameters. Note that the :func:`torch.view_as_real`
+equivalent will convert a complex tensor to a real tensor with shape :math:`(..., 2)`, which is distinct from splitting
```
I think we are over-indexing this explanation on that very small details.
I would split this to a second paragraph that covers:
- There can be numerical discrepancies between the two (complex vs view as real-ed) just like capturable/foreach/etc Linking to https://pytorch.org/docs/stable/notes/numerical_accuracy.html for details
- The note about one Tensor vs two Tensors. I think it's worth mentioning that this will only make a different for optimizers that do per-Tensor reductions (which we don't have any of) and we don't define this behavior yet, open an issue if you have a use case that needs this behavior to be defined.

janeyx99 committed:
Update on "Document complex optimizer semantic behavior"

<img width="975" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/7e7b87ca-ac62-4037-9173-3a805ac8c402">




[ghstack-poisoned]

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -129,9 +129,38 @@ Autograd
+`p.real` and `p.imag` pieces separately, the implementation details are not precisely that. Note that the
+:func:`torch.view_as_real` equivalent will convert a complex tensor to a real tensor with shape :math:`(..., 2)`,
+whereas splitting a complex tensor into two tensors is more like :math:`(2, ...)`. This distinction has no impact on
+pointwise optimizers (like AdamW) but will cause slight discrepancy in optimizers that do per-Tensor reductions. We
```
I'm not sure here. That would be more 2 Tensors of size (...)
And then you have two things:
- it has small differences for optimizers with global reduction (like lbfgs)
- it will give completely different results for optimizers that do per-param reduction (we don't have any of these)

janeyx99 replied:
okay whew this makes so much more sense i was curious what you meant before

janeyx99 committed:
Update on "Document complex optimizer semantic behavior"

<img width="975" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/7e7b87ca-ac62-4037-9173-3a805ac8c402">




[ghstack-poisoned]

albanD reviewed (approved):
Sounds good! Thanks for the clarification!

janeyx99 commented:
@pytorchbot merge
